### PR TITLE
Check for module variables in kernels to be transformed (for #315)

### PR DESCRIPTION
**Pull request by arporter**
Uses the new PSyIR SymbolTable to check for global symbols in kernels that are the target of either a KernelInlineTrans or an ACCRoutineTrans.

### CODE REVIEW

**Comment by arporter** [comment]
Realised in writing the description of this test that this check for global data also needs to be a part of the OpenCL transformation.

**Comment by arporter** [comment]
For some reason CodeCov has stopped reporting but I've checked coverage manually.
Ready for first review, probably from @sergisiso or @rupertford.

**Comment by rupertford** [comment]
Back to @arporter

File: examples/gocean/eg4/README [hunk 1]
```diff
@@ -63,17 +63,16 @@ of kernels accessing data/routines via module use statements:
...
-In addition, although the kernels named in the Invoke have been transformed,
-the other kernels that they then call have not and this means that they
-won't be compiled for execution on the OpenACC device. Support for recursive
-kernel transformation is the subject of Issue #342.
+causes PSyclone to raise a TransformationError.  This is because the
+data_mod::gravity variable read by `kern_use_var` is not available on
+the remote device.  This problem is the subject of Issue #323 which
+proposes to extend PSyclone to convert any module data accessed by a
+kernel into a kernel argument.  In addition, although the kernels
+named in the Invoke have been transformed, the other kernels that they
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg4/README]: Well spotted. Done.

File: examples/gocean/eg4/README [hunk 2]
```diff
@@ -63,17 +63,16 @@ of kernels accessing data/routines via module use statements:
...
-convert any module data accessed by a kernel into a kernel argument.
-In addition, although the kernels named in the Invoke have been transformed,
-the other kernels that they then call have not and this means that they
-won't be compiled for execution on the OpenACC device. Support for recursive
-kernel transformation is the subject of Issue #342.
+causes PSyclone to raise a TransformationError.  This is because the
+data_mod::gravity variable read by `kern_use_var` is not available on
+the remote device.  This problem is the subject of Issue #323 which
+proposes to extend PSyclone to convert any module data accessed by a
+kernel into a kernel argument.  In addition, although the kernels
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg4/README]: Latex style. Fixed.

File: examples/gocean/eg4/README [hunk 3]
```diff
@@ -63,17 +63,16 @@ of kernels accessing data/routines via module use statements:
...
-is *not* a valid OpenACC application. This is because the data_mod::gravity
-variable read by `kern_use_var` is not available on the remote device.
-This problem is the subject of Issue #323 which proposes to extend PSyclone to
-convert any module data accessed by a kernel into a kernel argument.
-In addition, although the kernels named in the Invoke have been transformed,
-the other kernels that they then call have not and this means that they
-won't be compiled for execution on the OpenACC device. Support for recursive
-kernel transformation is the subject of Issue #342.
+causes PSyclone to raise a TransformationError.  This is because the
+data_mod::gravity variable read by `kern_use_var` is not available on
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg4/README]: I don't think so - I've removed it.

File: src/psyclone/tests/conftest.py [hunk 4]
```diff
@@ -140,3 +140,12 @@ def parser():
     '''
     from fparser.two.parser import ParserFactory
     return ParserFactory().create()
+
+
+@pytest.fixture
+def kernel_output_dir(tmpdir, monkeypatch):
+    '''Sets the Psyclone _kernel_output_dir Config parameter to tmpdir.'''
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/conftest.py]: Done.

File: src/psyclone/tests/kernel_transformation_test.py [hunk 5]
```diff
@@ -94,6 +94,36 @@ def test_accroutine_err(monkeypatch):
...
+    # Attempt to transform the second (time_smooth) kernel which uses
+    # a variable that is defined in the enclosing module.
+    with pytest.raises(TransformationError) as err:
+        _ = rtrans.apply(kernels[0])
+    assert ("'kernel_with_use_code' contains the following symbols with "
+            "'global' scope: ['rdt']. PSyclone cannot currently transform "
+            in str(err))
+
+
+def test_accroutine_module_use():
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/kernel_transformation_test.py]: Oh yes, don't know what happened there - deleted one of them.

File: src/psyclone/tests/test_files/dynamo0p1/algorithm/testkern.F90 [hunk 6]
```diff
@@ -1,10 +1,35 @@
...
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. W. Ford, STFC Daresbury Lab
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p1/algorithm/testkern.F90]: Done.

File: src/psyclone/tests/test_files/dynamo0p3/ru_kernel_mod.f90 [hunk 7]
```diff
@@ -78,6 +78,7 @@ subroutine ru_code(nlayers, field_1_w2, field_2_w3, iscalar_3, rscalar_4, field_
       INTEGER, intent(in) :: undf_w3
       INTEGER, intent(in) :: ndf_w0
       INTEGER, intent(in) :: undf_w0
+      INTEGER, intent(in) :: np_xy, np_z
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/ru_kernel_mod.f90]: Done.

File: src/psyclone/transformations.py [hunk 8]
```diff
@@ -51,6 +51,34 @@
 VALID_OMP_SCHEDULES = ["runtime", "static", "dynamic", "guided", "auto"]
 
 
+def check_intergrid(node):
+    '''
+    Utility function to check that the supplied node does not have
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py [hunk 9]
```diff
@@ -51,6 +51,34 @@
...
+    an intergrid kernel as a child.
+
+    This is used to ensure that we reject any attempt to apply
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py [hunk 10]
```diff
@@ -51,6 +51,34 @@
...
+    loop-fusion and redundant-computation transformations to loops containing
+    inter-grid kernels (since support for those is not yet implemented).
+
+    :param node: The PSyIR node to check.
+    :type node: :py:class:`psyGen.Node`
+
+    :raises TransformationError: if the supplied node has an inter-grid \
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: I disagree. If you look at the generated documentation then this is formatted something like:

Screenshot_20190716_101120

File: src/psyclone/transformations.py [hunk 11]
```diff
@@ -51,6 +51,34 @@
...
+    '''
+    if not node.children:
+        return
+    from psyclone.dynamo0p3 import DynKern
+    child_kernels = node.walk(node.children, DynKern)
+    for kern in child_kernels:
+        if kern.is_intergrid:
+            raise TransformationError(
+                "This Transformation cannot currently be applied to nodes "
+                "which have inter-grid kernels as children and {0} is such a "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py [hunk 12]
```diff
@@ -143,34 +171,50 @@ def _validate(self, node_list):
...
+    Base class for all Kernel transformations.
 
-    This is used to ensure that we reject any attempt to apply
-    transformations to loops containing inter-grid kernels. (This restriction
-    will be lifted in Issue #134 and this routine can then be removed.)
+    '''
+    @staticmethod
+    def validate(kern):
+        '''
+        Checks that the supplied node is a Kernel and that we can construct
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py [hunk 13]
```diff
@@ -2983,6 +3032,21 @@ def validate(self, kern):
                                       "it will be module-inlined.".
                                       format(kern.name))
 
+        # Perform general validation checks. In particular this checks that
+        # we can construct the PSyIR of the kernel body.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py [hunk 14]
```diff
@@ -2983,6 +3032,21 @@ def validate(self, kern):
...
+
+        # Check that the kernel does not access any data via a module 'use'
+        # statement
+        sched = kern.get_kernel_schedule()
+        global_symbols = sched.symbol_table.global_symbols
+        if global_symbols:
+            raise TransformationError(
+                "The Symbol Table for kernel '{0}' contains the following "
+                "symbols with 'global' scope: {1}. PSyclone cannot currently"
+                " transform kernels that access data not passed by argument.".
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/psyGen.py [hunk 15]
```diff
@@ -262,6 +262,21 @@ def __str__(self):
         return str(self.value)
 
 
+class SymbolError(Exception):
+    '''
+    PSyclone-specific exception for use with errors relating to the SymbolTable
+    in the PSyIR.
+
+    :param str value: the message associated with the error.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: See earlier :-)

File: src/psyclone/tests/psyGen_test.py [hunk 16]
```diff
@@ -4430,6 +4430,7 @@ def test_fparser2astprocessor_handling_name(f2008_parser):
     '''
     from fparser.common.readfortran import FortranStringReader
     from fparser.two.Fortran2003 import Execution_Part
+    from psyclone.psyGen import SymbolError
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Good point, done.

File: src/psyclone/transformations.py [hunk 17]
```diff
@@ -1279,18 +1322,12 @@ def validate(self, node, inline):
...
         :raises TransformationError: if the supplied kernel has itself been \
                                      transformed (Issue #229).
         '''
-        # check node is a kernel
-        from psyclone.psyGen import Kern
-        if not isinstance(node, Kern):
-            raise TransformationError(
-                "Error in KernelModuleInline transformation. The node is not "
-                "a Kernel")
+        KernelTrans.validate(node)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: Yes and done.

File: src/psyclone/transformations.py [hunk 18]
```diff
@@ -143,34 +171,50 @@ def _validate(self, node_list):
...
-    will be lifted in Issue #134 and this routine can then be removed.)
+    '''
+    @staticmethod
+    def validate(kern):
+        '''
+        Checks that the supplied node is a Kernel and that we can construct
+        the PSyIR of its contents.
 
-    # TODO remove this routine once #134 is complete.
+        :param kern: the kernel which is the target of the transformation.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: See earlier.

File: src/psyclone/transformations.py [hunk 19]
```diff
@@ -143,34 +171,50 @@ def _validate(self, node_list):
...
+        :type kern: :py:class:`psyclone.psyGen.Kern` or sub-class.
 
-    :param node: The PSyIR node to check.
-    :type node: :py:class:`psyGen.Node`
+        :raises TransformationError: if the target node is not a sub-class of \
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/transformations.py]: See earlier.

**Comment by arporter** [comment]
I've added a bit to the Developers' Guide as requested. Ready for second review.

**Comment by rupertford** [comment]
Back to @arporter

**Comment by arporter** [comment]
Conflicts fixed. Ready for next review from @rupertford.

File: src/psyclone/transformations.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: I agree, as discussed and now captured in the wiki.

File: src/psyclone/transformations.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: This doesn't appear to have been modified.

File: src/psyclone/transformations.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: Thanks, but there is now a missing "the" ... checks that the PSyIR ...

File: src/psyclone/psyGen.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyGen.py]: Agreed.

File: src/psyclone/transformations.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: Agreed.

File: src/psyclone/transformations.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: Agreed.

File: src/psyclone/transformations.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: I've updated it as it was a minor change.

File: src/psyclone/transformations.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/transformations.py]: I modified it myself as it is a minor change. Note, I added "a" not "the".

**Comment by rupertford** [comment]
Very nearly there. Please check the reported missing coverage (one line) and bring the branch up-to-date.

**Comment by arporter** [comment]
That wasn't fun. I've ended up re-naming `testkern.F90` to `testkern_mod.F90` (as well as the module it contains) so that it obeys the GungHo naming conventions - it was causing test failures because my new/extended validation attempts to construct the PSyIR of the target kernel. I also hit the fact that, because I'd constructed the PSyIR, the kernel generation was (incorrectly) using the PSyIR Fortran backend. This was wrong in two ways: 1) that backend currently generates LFRic-specific code and 2) it wasn't the PSyIR representation that had been transformed. I've worked around this for the moment by removing the KernelSchedule that is constructed during validation. The proper solution is to ensure all kernel transformations work with the PSyIR rather than the fparser2 parse tree (#490).

If Travis and CodeCov are happy then I'll pass this back for review...

**Pull request merged**